File: C#/2020/day1/Day1.cs
using System;
using System.Linq;

namespace AOC
{
    class Program
    {
        public static System.Collections.Generic.IEnumerable<int[]> solution(string[] lines)
        {
            for (int i = 0; i < lines.Length - 2; i++)
            {
                for (int j = i; j < lines.Length - 1; j++)
                {
                    for (int y = j; y < lines.Length; y++)
                    {
                        yield return new[] { Int32.Parse(lines[i]), Int32.Parse(lines[j]), Int32.Parse(lines[y]) };
                    }
                    yield return new[] { Int32.Parse(lines[i]), Int32.Parse(lines[j]) };
                }
            }
        }
        static void Main(string[] args)
        {
            string[] input = System.IO.File.ReadAllLines(@"day1.txt");

            foreach (var pair in solution(input))
            {
                int result = pair.Sum();
                var wasFound = false;
                if (result == 2020)
                {
                    if (pair.Length == 2)
                    {
                        Console.WriteLine("part1: {0}", pair.Aggregate(1, (a, b) => a * b));
                        if (!wasFound) wasFound = true;
                        else break;
                    }
                    else if (pair.Length == 3)
                    {
                        Console.WriteLine("part2: {0}", pair.Aggregate(1, (a, b) => a * b));
                        if (!wasFound) wasFound = true;
                        else break;
                    }
                }
            }
        }
    }
}

File: Python/2020/day1/day1.py
from functools import reduce
from itertools import combinations


def fun(n):
    for x in combinations([int(x) for x in open("input.txt")], n):
        if sum(x) == 2020:
            return reduce(lambda a, b: a * b, x)


# print(fun(2))
# print(fun(3))


def one_line(n):
    return [reduce(lambda a, b: a * b, x) for x in combinations([int(x) for x in open("input.txt")], n) if sum(x) == 2020]


print(*one_line(2))
print(*one_line(3))
